src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+    real_type end_curve_length = step;
+
+    real_type h = ((hinitial > rel_tolerance() * step) && (hinitial < step))
+                      ? hinitial
+                      : step;
+    real_type h_threshold = shared_.epsilon_step * step;
+
+    // Output with the next good step
+    FieldOutput output;
+
+    // Performance integration
+    bool         succeeded       = false;
+    real_type    curve_length    = 0;
+    unsigned int remaining_steps = shared_.max_nsteps;
+
+    do
+    {
+        output = this->integrate_step(h, *state);
+
+        curve_length += output.step_taken;
+
+        if (h < h_threshold || curve_length >= end_curve_length)
+        {
+            succeeded = true;
+        }
+        else
+        {
+            h = std::fmax(output.value, shared_.minimum_step);
+            if (curve_length + h > end_curve_length)
+            {
+                h = end_curve_length - curve_length;
+            }
+        }
+        *state = output.state;
+    } while (!succeeded && --remaining_steps > 0);
+
+    CELER_ASSERT(succeeded);
+
+    return curve_length;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function for accurate_advance: advance for a given step and
+ * evaluate the next predicted step.
+ */
+CELER_FUNCTION auto
+FieldDriver::integrate_step(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a next proposed step
+    FieldOutput output;
+
+    if (step > shared_.minimum_step)
+    {
+        output = this->one_good_step(step, state);
+    }
+    else
+    {
+        // Do an integration step for a small step (a.k.a quick advance)
+        StepperResult result = stepper_(step, state);
+
+        // Update position and momentum
+        output.state = result.end_state;
+
+        real_type dyerr = truncation_error(
+            step, shared_.epsilon_rel_max, state, result.err_state);
+        output.step_taken = step;
+
+        // Compute a proposed new step
+        CELER_ASSERT(output.step_taken != 0);
+        output.value
+            = this->new_step_size(step, dyerr / (step * shared_.epsilon_step));
+    }
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Advance within a relative truncation error and estimate a good step size
+ * for the next integration.
+ */
+CELER_FUNCTION auto
+FieldDriver::one_good_step(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a proposed next step
+    FieldOutput output;
+
+    // Perform integration for adaptive step control with the trunction error
+    bool          condition       = false;
+    unsigned int  remaining_steps = shared_.max_nsteps;
+    real_type     errmax2 = celeritas::numeric_limits<real_type>::max();
+    StepperResult result;
+
+    do
+    {
+        result  = stepper_(step, state);
+        errmax2 = truncation_error(
```

Comment:
truncation_error and distance_chord are from FieldUtils.hh (just under the celeritas name space).  I thought about adding detail:: namespace, but postpone to create detail directory and the name space, since there is only one file
involved (FieldUtils), until other files are populated.

Reply:
Fair enough.